File: cameo-api-cpp/src/coms/impl/RequesterImpl.cpp
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

#include "RequesterImpl.h"

#include "Application.h"
#include "Serializer.h"
#include "JSON.h"
#include "../../base/impl/RequestSocketImpl.h"
#include "../../base/impl/ContextImpl.h"
#include "../../base/Messages.h"
#include <sstream>

namespace cameo {
namespace coms {

const std::string RequesterImpl::REQUESTER_PREFIX = "req.";
std::mutex RequesterImpl::m_mutex;
int RequesterImpl::m_requesterCounter = 0;

RequesterImpl::RequesterImpl(const Endpoint& endpoint, int requesterPort, int responderPort, const std::string& name, int responderId, int requesterId) :
	m_requesterPort(requesterPort),
	m_name(name),
	m_responderId(responderId),
	m_requesterId(requesterId),
	m_canceled(false) {

	// Create the request socket.
	m_requestSocket = application::This::getCom().createRequestSocket(endpoint.withPort(responderPort).toString());

	// Create a socket REP.
	ContextImpl* contextImpl = dynamic_cast<ContextImpl *>(application::This::getCom().getContext());
	m_repSocket.reset(new zmq::socket_t(contextImpl->m_context, ZMQ_REP));
	std::stringstream reqEndpoint;
	reqEndpoint << "tcp://*:" << m_requesterPort;

	m_repSocket->bind(reqEndpoint.str().c_str());
}

RequesterImpl::~RequesterImpl() {
	terminate();
}

int RequesterImpl::newRequesterId() {

	std::lock_guard<std::mutex> lock(m_mutex);
	m_requesterCounter++;

	return m_requesterCounter;
}

std::string RequesterImpl::getRequesterPortName(const std::string& name, int responderId, int requesterId) {

	std::stringstream requesterPortName;
	requesterPortName << REQUESTER_PREFIX << name << "." << responderId << "." << requesterId;

	return requesterPortName.str();
}

WaitingImpl * RequesterImpl::waiting() {
	return new GenericWaitingImpl(std::bind(&RequesterImpl::cancel, this));
}

void RequesterImpl::sendBinary(const std::string& requestData) {

	json::StringObject request;
	request.pushKey(message::TYPE);
	request.pushInt(message::REQUEST);

	request.pushKey(message::Request::APPLICATION_NAME);
	request.pushString(application::This::getName());

	request.pushKey(message::Request::APPLICATION_ID);
	request.pushInt(application::This::getId());

	request.pushKey(message::Request::SERVER_URL);
	request.pushString(application::This::getEndpoint().getProtocol() + "://" + application::This::getEndpoint().getAddress());

	request.pushKey(message::Request::SERVER_PORT);
	request.pushInt(application::This::getEndpoint().getPort());

	request.pushKey(message::Request::REQUESTER_PORT);
	request.pushInt(m_requesterPort);

	m_requestSocket->request(request.toString(), requestData);
}

void RequesterImpl::send(const std::string& requestData) {

	// encode the data
	std::string result;
	serialize(requestData, result);
	sendBinary(result);
}

void RequesterImpl::sendTwoBinaryParts(const std::string& requestData1, const std::string& requestData2) {

	json::StringObject request;
	request.pushKey(message::TYPE);
	request.pushInt(message::REQUEST);

	request.pushKey(message::Request::APPLICATION_NAME);
	request.pushString(application::This::getName());

	request.pushKey(message::Request::APPLICATION_ID);
	request.pushInt(application::This::getId());

	request.pushKey(message::Request::SERVER_URL);
	request.pushString(application::This::getEndpoint().getProtocol() + "://" + application::This::getEndpoint().getAddress());

	request.pushKey(message::Request::SERVER_PORT);
	request.pushInt(application::This::getEndpoint().getPort());

	request.pushKey(message::Request::REQUESTER_PORT);
	request.pushInt(m_requesterPort);

	m_requestSocket->request(request.toString(), requestData1, requestData2);
}

std::optional<std::string> RequesterImpl::receiveBinary() {

	if (m_canceled) {
		return {};
	}

	std::unique_ptr<zmq::message_t> message(new zmq::message_t);
	m_repSocket->recv(message.get(), 0);

	// Get the JSON request.
	json::Object request;
	json::parse(request, message.get());

	int type = request[message::TYPE].GetInt();

	if (type == message::CANCEL) {
		m_canceled = true;
		return {};
	}

	std::optional<std::string> result;

	if (type == message::RESPONSE) {
		// Get the second part for the message.
		message.reset(new zmq::message_t);
		m_repSocket->recv(message.get(), 0);
		result = std::string(message->data<char>(), message->size());
	}

	// Create the reply.
	std::string data = "OK";
	size_t size = data.length();
	std::unique_ptr<zmq::message_t> reply(new zmq::message_t(size));
	memcpy(reply->data(), data.c_str(), size);

	m_repSocket->send(*reply);

	return result;
}

std::optional<std::string> RequesterImpl::receive() {
	return receiveBinary();
}

void RequesterImpl::cancel() {

	json::StringObject request;
	request.pushKey(message::TYPE);
	request.pushInt(message::CANCEL);

	// Create a request socket only for the request.
	std::unique_ptr<RequestSocketImpl> requestSocket = application::This::getCom().createRequestSocket(application::This::getEndpoint().withPort(m_requesterPort).toString());
	requestSocket->request(request.toString());
}

void RequesterImpl::terminate() {

	if (m_repSocket.get() != nullptr) {
		m_repSocket.reset(nullptr);

		application::This::getCom().removePort(getRequesterPortName(m_name, m_responderId, m_requesterId));
	}

	m_requestSocket.reset();
}

}
}

File: cameo-api/src/main/java/fr/ill/ics/cameo/coms/impl/RequesterImpl.java
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

package fr.ill.ics.cameo.coms.impl;

import java.util.concurrent.atomic.AtomicInteger;

import org.json.simple.JSONObject;

import fr.ill.ics.cameo.Zmq;
import fr.ill.ics.cameo.base.Application.This;
import fr.ill.ics.cameo.base.impl.ContextImpl;
import fr.ill.ics.cameo.base.impl.RequestSocket;
import fr.ill.ics.cameo.base.impl.ThisImpl;
import fr.ill.ics.cameo.messages.JSON;
import fr.ill.ics.cameo.messages.Messages;
import fr.ill.ics.cameo.strings.Endpoint;

public class RequesterImpl {

	public static final String REQUESTER_PREFIX = "req.";

	private ThisImpl application;
	private int requesterPort;
	private String name;
	private int responderId;
	
	// Need for a unique id per Application instance.
	private int requesterId;
	private static AtomicInteger requesterCounter = new AtomicInteger();
	
	private Zmq.Context context;
	private Zmq.Socket requester;
	private RequestSocket requestSocket;
	
	private boolean canceled = false;
	private RequesterWaitingImpl waiting = new RequesterWaitingImpl(this);
		
	public RequesterImpl(Endpoint endpoint, int requesterPort, int responderPort, String name, int responderId, int requesterId) {
		this.requesterPort = requesterPort;
		this.name = name;
		this.responderId = responderId;
		this.requesterId = requesterId;
		this.context = ((ContextImpl)This.getCom().getContext()).getContext();

		// Create the REQ socket.
		String responderEndpoint = endpoint.withPort(responderPort).toString();
		requestSocket = This.getCom().createRequestSocket(responderEndpoint);
		
		// Create the REP socket.
		requester = context.createSocket(Zmq.REP);
		requester.bind("tcp://*:" + requesterPort);
		
		waiting.add();
	}
	
	public String getName() {
		return name;
	}
	
	public static int newRequesterId() {
		return requesterCounter.incrementAndGet();
	}
	
	public static String getRequesterPortName(String name, int responderId, int requesterId) {
		return REQUESTER_PREFIX + name + "." + responderId + "." + requesterId;
	}
		
	public void send(byte[] requestData) {
		
		JSONObject request = new JSONObject();
		request.put(Messages.TYPE, Messages.REQUEST);
		request.put(Messages.Request.APPLICATION_NAME, This.getName());
		request.put(Messages.Request.APPLICATION_ID, This.getId());
		request.put(Messages.Request.SERVER_URL, This.getEndpoint().getProtocol() + "://" + This.getEndpoint().getAddress());
		request.put(Messages.Request.SERVER_PORT, This.getEndpoint().getPort());
		request.put(Messages.Request.REQUESTER_PORT, requesterPort);
		
		Zmq.Msg message = new Zmq.Msg();
		message.add(Messages.serialize(request));
		
		// Set request in the next frame.
		message.add(requestData);
		
		requestSocket.request(message);
	}
	
	public void send(String request) {
		send(Messages.serialize(request));
	}
	
	public void sendTwoParts(byte[] requestData1, byte[] requestData2) {
		
		JSONObject request = new JSONObject();
		request.put(Messages.TYPE, Messages.REQUEST);
		request.put(Messages.Request.APPLICATION_NAME, This.getName());
		request.put(Messages.Request.APPLICATION_ID, This.getId());
		request.put(Messages.Request.SERVER_URL, This.getEndpoint().getProtocol() + "://" + This.getEndpoint().getAddress());
		request.put(Messages.Request.SERVER_PORT, This.getEndpoint().getPort());
		request.put(Messages.Request.REQUESTER_PORT, requesterPort);
		
		Zmq.Msg message = new Zmq.Msg();
		message.add(Messages.serialize(request));
		
		// Set request1 and request2 in the next frames.
		message.add(requestData1);
		message.add(requestData2);
		
		requestSocket.request(message);
	}

	public byte[] receive() {
		
		Zmq.Msg message = null;
		
		try {
			message = Zmq.Msg.recvMsg(requester);

			if (message == null) {
				return null;
			}
			
			// Get the JSON request object.
			JSONObject request = This.getCom().parse(message.getFirstData());
			
			// Get the type.
			long type = JSON.getLong(request, Messages.TYPE);
						
			if (type == Messages.RESPONSE) {
				return message.getLastData();
			}
			else if (type == Messages.CANCEL) {
				canceled = true;
				return null;
			}
			else {
				return null;
			}
		}
		finally {
			if (message != null) {
				message.destroy();
			}
			
			// Send to the responder
			Zmq.Msg reply = new Zmq.Msg();
			reply.add("OK");
			reply.send(requester);
		}
	}

	public String receiveString() {
		return Messages.parseString(receive());
	}
	
	public void cancel() {
		
		Endpoint endpoint = This.getEndpoint().withPort(requesterPort);

		JSONObject request = new JSONObject();
		request.put(Messages.TYPE, Messages.CANCEL);
		
		// Create the request socket. We can create it here because it should be called only once.
		RequestSocket requestSocket = This.getCom().createRequestSocket(endpoint.toString());
		requestSocket.request(request);
		
		// Terminate the socket.
		requestSocket.terminate();
	}

	public boolean isCanceled() {
		return canceled;
	}
	
	public void terminate() {
		
		waiting.remove();
		
		// Terminate the request socket.
		requestSocket.terminate();
		
		context.destroySocket(requester);
		
		try {
			This.getCom().removePort(getRequesterPortName(name, responderId, requesterId));
			
		} catch (Exception e) {
			System.err.println("Cannot terminate requester: " + e.getMessage());
		}
	}
	
	@Override
	public String toString() {
		return REQUESTER_PREFIX + name + "." + requesterId + ":" + This.getName() + "." + This.getId() + "@" + This.getEndpoint();
	}
	
}